src/instruments.rs

```diff
@@ -1,133 +1,314 @@
+            let clt_git_filepath = Path::new("/Library/Developer/CommandLineTools/usr/bin/git");
+            if clt_git_filepath.exists() {
+                return Ok(XcodeInstruments::XcTrace);
+            }
+        } else {
+            let instruments_app_filepath = Path::new("/usr/bin/instruments");
+            if instruments_app_filepath.exists() {
+                return Ok(XcodeInstruments::InstrumentsBinary);
+            }
+        }
         Err(anyhow!(
-            "/usr/bin/instruments does not exist. \
-             Please install the Xcode Command Line Tools."
+            "Xcode Instruments is not installed. Please install the Xcode Command Line Tools."
         ))
     }
+
+    /// Return a string listing available Instruments Templates.
+    ///
+    /// The typical output of the underlying command is:
+    ///
+    /// ```
+    /// == Standard Templates ==
+    /// Activity Monitor
+    /// Allocations
+    /// Animation Hitches
+    /// App Launch
+    /// Core Data
+    /// Counters
+    /// Energy Log
+    /// File Activity
+    /// Game Performance
+    /// Leaks
+    /// Logging
+    /// Metal System Trace
+    /// Network
+    /// SceneKit
+    /// SwiftUI
+    /// System Trace
+    /// Time Profiler
+    /// Zombies
+    ///
+    /// == Custom Templates ==
+    /// MyTemplate
+    /// ```
+    ///
+    /// The custom templates only appears if you have custom templates.
+    pub(crate) fn available_templates(&self) -> Result<TemplateCatalog> {
+        let Output { status, stdout, .. } = match self {
+            XcodeInstruments::XcTrace => {
+                Command::new("xcrun").args(&["xctrace", "list", "templates"]).output()?
+            }
+            XcodeInstruments::InstrumentsBinary => {
+                Command::new("instruments").args(&["-s", "templates"]).output()?
+            }
+        };
+
+        if !status.success() {
+            return Err(anyhow!(
+                "Could not list templates. Please check your Xcode Instruments installation."
+            ));
+        }
+
+        let templates_str = String::from_utf8(stdout)?;
+        let mut templates_iter = templates_str.lines();
+
+        let standard_templates = templates_iter
+            .by_ref()
+            .skip(1)
+            .map(|line| line.trim())
+            .take_while(|line| !line.starts_with('=') && !line.is_empty())
+            .map(|line| line.into())
+            .collect::<Vec<_>>();
+
+        if standard_templates.is_empty() {
+            return Err(anyhow!(
+                "No available templates. Please check your Xcode Instruments installation."
+            ));
+        }
+
+        let custom_templates = templates_iter
+            .map(|line| line.trim())
+            .skip_while(|line| line.starts_with('=') || line.is_empty())
+            .map(|line| line.into())
+            .collect::<Vec<_>>();
+
+        let catalog = TemplateCatalog { standard_templates, custom_templates };
+        Ok(catalog)
```

Comment:
okay, after some testing I have some concerns about this method:

- the output of `xcrun xctrace list templates` and `instruments -s templates` differs. The former looks like what you're checking for, but the latter looks like:

```
$ instruments -s templates

Known Templates:
"Activity Monitor"
"Allocations"
"Animation Hitches"
"App Launch"
"Blank"
"Core Data"
"Counters"
"Energy Log"
"File Activity"
"Game Performance"
"Leaks"
"Logging"
"Metal System Trace"
"Network"
"SceneKit"
"SwiftUI"
"System Trace"
"Time Profiler"
"Zombies"
"~/Library/Application Support/Instruments/Templates/MyTemplate.tracetemplate"
```

In additon, on my laptop (macOS `10.15.7`, Xcode `12.4`) The output of `xcrun xtrace list templates` is printed to *stderr*, not *stdout*. I guess this changed at some point?

In any case, it looks like these tools will, at the very least, require separate parsing code; and in the case of `xcrun xctrace` if stdout is empty but stderr is not, I would use stderr instead of stdout.

Additionally, I'm curious if there is any practical different between the standard and custom templates? Do we treat them differently in any way?

Reply:
For the output of `/usr/bin/instruments`: Argh, absolutely true, sorry about that. I'll soon add the missing parsing logic.

Yes, the output of `xcrun xctrace` changed to stdout, at some point (macOS 11.0?) I saw it on a forum, but I forgot to change that, sorry again, and thanks for your throughout testing. I guess that's an easy fix. I'm happy to apply the fix you described, but we might have to iterate as I have no way to test it on 10.15 anymore.

Yes you're right, the 2 kinds of templates are treated identically. The separation is just for bookeeping, but that's all.

I went for supporting custom templates for completeness, and also because [custom instruments](https://developer.apple.com/wwdc18/410) are pretty cool.